les/server.go

```diff
@@ -140,179 +175,179 @@ func (s *LesServer) Stop() {
-	return l
+	sectionHead, chtRoot, bloomTrieRoot := s.getCheckpoint(count - 1)
+	return count - 1, sectionHead, chtRoot, bloomTrieRoot
 }
 
-type requestCostStats struct {
-	lock  sync.RWMutex
-	db    ethdb.Database
-	stats map[uint64]*linReg
-}
+// getCheckpoint returns a set of post-processed trie roots (CHT and BloomTrie)
+// associated with the appropriate head hash by specific section index.
+func (s *LesServer) getCheckpoint(index uint64) (common.Hash, common.Hash, common.Hash) {
+	// convert last LES/2 section index back to LES/1 index for chtIndexer.SectionHead
+	latest := (index+1)*(light.CHTFrequencyClient/light.CHTFrequencyServer) - 1
 
-type requestCostStatsRlp []struct {
-	MsgCode uint64
-	Data    []byte
+	sectionHead := s.chtIndexer.SectionHead(latest)
+	chtRoot := light.GetChtRoot(s.protocolManager.chainDb, latest, sectionHead)
+	bloomTrieRoot := light.GetBloomTrieRoot(s.protocolManager.chainDb, index, sectionHead)
+	return sectionHead, chtRoot, bloomTrieRoot
 }
 
-var rcStatsKey = []byte("_requestCostStats")
-
-func newCostStats(db ethdb.Database) *requestCostStats {
-	stats := make(map[uint64]*linReg)
-	for _, code := range reqList {
-		stats[code] = &linReg{cnt: 100}
+// checkpointLoop starts a standalone goroutine to watch new checkpoint events and updates local's stable checkpoint.
+func (s *LesServer) checkpointLoop(checkpoint *light.TrustedCheckpoint) (err error) {
+	var (
+		eventCh      = make(chan *contract.ContractNewCheckpointEvent)
+		headCh       = make(chan core.ChainHeadEvent, SubscribeChainHeadEvent)
+		announcement = make(map[uint64]common.Hash)
+	)
+	eventSub, err := s.registrar.WatchNewCheckpointEvent(eventCh)
+	if err != nil {
+		return err
+	}
+	headSub := s.backend.SubscribeChainHeadEvent(headCh)
+	if headSub == nil {
+		eventSub.Unsubscribe()
+		return errors.New("subscribe head event failed")
 	}
 
-	if db != nil {
-		data, err := db.Get(rcStatsKey)
-		var statsRlp requestCostStatsRlp
-		if err == nil {
-			err = rlp.DecodeBytes(data, &statsRlp)
-		}
-		if err == nil {
-			for _, r := range statsRlp {
-				if stats[r.MsgCode] != nil {
-					if l := linRegFromBytes(r.Data); l != nil {
-						stats[r.MsgCode] = l
-					}
+	ticker := time.NewTicker(5 * time.Minute)
+	defer func() {
+		eventSub.Unsubscribe()
+		headSub.Unsubscribe()
+		ticker.Stop()
+	}()
+
+	for {
+		select {
+		case event := <-eventCh:
+			if event == nil {
+				// This should never happen.
+				log.Info("Ignore empty checkpoint event")
+				continue
+			}
+			// Note several events have same index may be received because of chain reorg and
+			// the modification of the latest checkpoint.
+			if checkpoint == nil || event.Index.Uint64() > checkpoint.SectionIdx {
+				log.Info("Receive new checkpoint event", "section", event.Index, "hash", common.Hash(event.CheckpointHash).Hex(),
+					"grantor", event.Grantor.Hex())
+				announcement[event.Index.Uint64()] = common.Hash(event.CheckpointHash)
+			}
+		case head := <-headCh:
```

Comment:
This is not quite right. You are triggering when there are enough blocks for a new section to be calculated but then you immediately assume that the section has been processed and you can retrieve it from the chain indexer. The chain indexer needs a bit of time to do the processing. You should act on finished sections, not new heads. Unfortunately the indexer does not have an event feed feature yet but you can add one (you should send events to subscribed listeners from func (c *ChainIndexer) setValidSections when storedSections is changed).

Reply:
If at that time(announcement received, current block number > section height + 256) cht root and bloom trie root have not been calculated, the generated checkpoint will mismatch the hash attached in announcement. So that server will try to generate a stable checkpoint and match the hash in the next few blocks until indexer finishes calculation.

And for order should be:
- indexer calculates and generates unstable checkpoint 
- admin fetches and register the unstable checkpoint
- server receives checkpoint announcement
- persist stable checkpoint

If we subscribe events from indexer, the logic will only be triggered every 32768 blocks. And for the usual case, `new section event` will be received early than `unstable checkpoint announcement`. So that we can not update stable checkpoint asap